cpp/src/io/parquet/bloom_filter_reader.cu

```diff
@@ -0,0 +1,680 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "arrow_filter_policy.cuh"
+#include "compact_protocol_reader.hpp"
+#include "io/parquet/parquet.hpp"
+#include "reader_impl_helpers.hpp"
+
+#include <cudf/ast/detail/expression_transformer.hpp>
+#include <cudf/ast/detail/operators.hpp>
+#include <cudf/ast/expressions.hpp>
+#include <cudf/detail/cuco_helpers.hpp>
+#include <cudf/detail/transform.hpp>
+#include <cudf/detail/utilities/logger.hpp>
+#include <cudf/hashing/detail/xxhash_64.cuh>
+#include <cudf/utilities/span.hpp>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/utilities/type_checks.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_buffer.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <cuco/bloom_filter_ref.cuh>
+#include <thrust/iterator/counting_iterator.h>
+
+#include <future>
+#include <numeric>
+#include <optional>
+
+namespace cudf::io::parquet::detail {
+namespace {
+
+/**
+ * @brief Converts bloom filter membership results (for each column chunk) to a device column.
+ *
+ */
+struct bloom_filter_caster {
+  cudf::device_span<cudf::device_span<cuda::std::byte> const> bloom_filter_spans;
+  host_span<Type const> parquet_types;
+  size_t total_row_groups;
+  size_t num_equality_columns;
+
+  enum class is_int96_timestamp : bool { YES, NO };
+
+  template <typename T, is_int96_timestamp IS_INT96_TIMESTAMP = is_int96_timestamp::NO>
+  std::unique_ptr<cudf::column> query_bloom_filter(cudf::size_type equality_col_idx,
+                                                   cudf::data_type dtype,
+                                                   ast::literal const* const literal,
+                                                   rmm::cuda_stream_view stream,
+                                                   rmm::device_async_resource_ref mr) const
+  {
+    using key_type    = T;
+    using policy_type = cuco::arrow_filter_policy<key_type, cudf::hashing::detail::XXHash_64>;
+    using word_type   = typename policy_type::word_type;
+
+    // Check if the literal has the same type as the predicate column
+    CUDF_EXPECTS(
+      cudf::have_same_types(cudf::column_view{dtype, 0, {}, {}, 0, 0, {}},
+                            cudf::column_view{literal->get_data_type(), 0, {}, {}, 0, 0, {}}),
+      "Mismatched predicate column and literal types");
+
+    // Filter properties
+    auto constexpr word_size       = sizeof(word_type);
+    auto constexpr words_per_block = policy_type::words_per_block;
+
+    rmm::device_buffer results{total_row_groups, stream, mr};
+
+    // Query literal in bloom filters from each column chunk (row group).
+    thrust::for_each(
+      rmm::exec_policy_nosync(stream),
+      thrust::counting_iterator<size_t>{0},
+      thrust::counting_iterator{total_row_groups},
```

Comment:
I think thrust::tabulate works here. Maybe make a bool span of `results` and then the call looks something like this
```
thrust::tabulate(rmm::exec_policy_nosync(stream),res_span.begin(), res_span.end(), <the 
 same lambda, just return val instead of setting results[row_group_idx]>)
 ```